tests/upgrade_path/test_upgrade_path.py

```diff
@@ -44,7 +41,75 @@ def upgrade_path_lists(request):
+    logger.info("Cleaning up previously installed images on DUT")
+    current_os_version = duthost.shell('sonic_installer list | grep Current | cut -f2 -d " "')['stdout']
+    duthost.shell("sonic_installer set_next_boot {}".format(current_os_version), module_ignore_errors=True)
+    duthost.shell("sonic_installer set-next-boot {}".format(current_os_version), module_ignore_errors=True)
+    duthost.shell("sonic_installer cleanup -y", module_ignore_errors=True)
+
+
+def setup_upgrade_test(duthost, localhost, from_image, to_image, tbinfo,
+                       upgrade_type, modify_reboot_script=None, allow_fail=False):
+    logger.info("Test upgrade path from {} to {}".format(from_image, to_image))
+    cleanup_prev_images(duthost)
+    # Install base image
+    logger.info("Installing {}".format(from_image))
+    try:
+        target_version = install_sonic(duthost, from_image, tbinfo)
+    except RunAnsibleModuleFail as err:
+        migration_err_regexp = r"Traceback.*migrate_sonic_packages.*SonicRuntimeException"
+        msg = err.results['msg'].replace('\n', '')
+        if re.search(migration_err_regexp, msg):
+            logger.info(
+                "Ignore the package migration error when downgrading to from_image")
+            target_version = duthost.shell(
+                "cat /tmp/downloaded-sonic-image-version")['stdout']
+        else:
+            raise err
+    # Remove old config_db before rebooting the DUT in case it is not successfully
+    # removed in install_sonic due to migration error
+    logger.info("Remove old config_db file if exists, to load minigraph from scratch")
+    if duthost.shell("ls /host/old_config/minigraph.xml", module_ignore_errors=True)['rc'] == 0:
+        duthost.shell("rm -f /host/old_config/config_db.json")
+    # Perform a cold reboot
+    logger.info("Cold reboot the DUT to make the base image as current")
+    # for 6100 devices, sometimes cold downgrade will not work, use soft-reboot here
+    reboot_type = 'soft' if "s6100" in duthost.facts["platform"] else 'cold'
+    reboot(duthost, localhost, reboot_type=reboot_type)
+    check_sonic_version(duthost, target_version)
+
+    # Install target image
+    logger.info("Upgrading to {}".format(to_image))
+    target_version = install_sonic(duthost, to_image, tbinfo)
+
+    if allow_fail and modify_reboot_script:
+        # add fail step to reboot script
+        modify_reboot_script(upgrade_type)
+
+
+@pytest.mark.device_type('vs')
+def test_double_upgrade_path(localhost, duthosts, ptfhost, rand_one_dut_hostname,
+                      nbrhosts, fanouthosts, tbinfo, restore_image,                     # noqa F811
+                      get_advanced_reboot, verify_dut_health, advanceboot_loganalyzer,  # noqa F811
+                      upgrade_path_lists):                                              # noqa F811
+    duthost = duthosts[rand_one_dut_hostname]
+    upgrade_type, from_image, to_image, _, enable_cpa = upgrade_path_lists
```

Comment:
What if `from_image` or `to_image` consist of multiple images? This refactor seems to break that scenario. Assuming this new testcase is meant to support the A->B->B use case, it's not clear to me how it is currently accomplishing that.

Reply:
Even today, if `from_image` or `to_image` contained multiple images, then there would be failures; this changeset doesn't change that. I'm considering renaming the options in the future to make it clear that only one image can be passed in into those arguments, but that would be a user-breaking change, and so would need to be done at the right time.

For A->B->B, there are only two images involved; A is the `from_image`, and B is the `to_image`. There would be a cold-reboot done to A, then warm-reboot to B, followed by another warm-reboot to B. This is what is tested in the `test_double_upgrade_path` function.